src/deploy/extensions/planner.ts

```diff
@@ -6,7 +6,30 @@ import { FirebaseError } from "../../error";
 import { getFirebaseProjectParams, substituteParams } from "../../extensions/extensionsHelper";
 import { logger } from "../../logger";
 import { readInstanceParam } from "../../extensions/manifest";
+import { ParamBindingOptions } from "../../extensions/paramHelper";
 
+/**
+ * Instance spec used by manifest.
+ *
+ * Params are passed in ParamBindingOptions so we know the param bindings for
+ * all environments user has configured.
+ *
+ * So far this is only used for writing to the manifest, but in the future
+ * we want to read manifest into this interface.
+ */
+export interface ManifestInstanceSpec {
```

Comment:
I think creating this type split is reasonable for now, but I'd prefer to join these into a single type in the future. They have a ton of overlap (and InstanceSpec has some useful caching functions to fetch extension & extensionVersion), and the only difference is params, so I'd rather we move the type split down to just params. Maybe something like:
```
export type ParamValue = string | ParamBindingOptions
export function resolveParamValues(p: ParamValue) {
// Either resolves the ParamBindingOption or returns the string value
}
```

Reply:
I explored in that direction but not exactly a union type, I decided to not touch `InstanceSpec` because it's used in too many places, and it'd be hard to untangle..

The reason why I ended up introducing another type is because I thought conceptually it made sense. `ManifestInstanceSpec` is the bare minimum spec that we pull from the manifest. But after hydration with ExtensionVersion, Extension, and collapsing the params , it turns into `DeploymentInstanceSpec`, something we use for deployment/emulator. So in the long term I'm leaning towards keeping both types but maybe make them more clearly separated? wdyt?